fix a corner case of torch.arange

yuguo68 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #80758

trying to fix https://github.com/pytorch/pytorch/issues/80589, see the two corner cases in the issue. 
added the two cases in unit tests and add device to the tests. 

Edit: the landed version only fixes the first issue in https://github.com/pytorch/pytorch/issues/80589

yuguo68 committed:
fix two corner cases of torch.arange

[ghstack-poisoned]

yuguo68 committed:
lint on "fix two corner cases of torch.arange"


trying to fix https://github.com/pytorch/pytorch/issues/80589, see the two corner cases in the issue. 
added the two cases in unit tests and add device to the tests. 

[ghstack-poisoned]

ezyang commented:
It would have been nice for the device parametrization to be separately stacked (no need to do now, I'll review as is)

ezyang commented on test/test_tensor_creation_ops.py:
```diff
@@ -2472,124 +2472,143 @@ def test_range_warning(self, device):
         # FloatTensor
-        res1 = torch.arange(0.6, 0.89, 0.1, out=torch.FloatTensor())
+        out = torch.tensor([], dtype=torch.float, device=device)
+        res1 = torch.arange(0.6, 0.89, 0.1, out=out)
```
good riddance, thx

ezyang commented on aten/src/ATen/native/RangeFactories.cpp:
```diff
@@ -143,12 +143,20 @@ Tensor& range_out(const Scalar& start, const Scalar& end, const Scalar& step, Te
     auto xend = end.to<accscalar_t>();
     auto xstep = step.to<accscalar_t>();
 
+    TORCH_CHECK(xstep > 0 || xstep < 0, "step must be nonzero");
```
Lol are there some floating point shenanigans why `xstep != 0` is invalid here

ezyang replied:
preexisting condition

ezyang commented on test/test_tensor_creation_ops.py:
```diff
@@ -2472,124 +2472,143 @@ def test_range_warning(self, device):
+        r = torch.arange(0, 6, 3, device=device)
+        self.assertEqual(r.min(), 0)
+        self.assertEqual(r.max(), 3)
+        self.assertEqual(r.numel(), 2)
```
this is the new test

ezyang commented on test/test_tensor_creation_ops.py:
```diff
@@ -2472,124 +2472,143 @@ def test_range_warning(self, device):
+        r = torch.arange(0, -5, -2, device=device)
+        self.assertEqual(r.min(), -4)
+        self.assertEqual(r.max(), 0)
+        self.assertEqual(r.numel(), 3)
```
and this too, the test here is so bad woof

ezyang commented:
onnx failure looks real

yuguo68 commented:
>

yuguo68 commented:
> onnx failure looks real

thanks for reviewing. Yes, it turns out in onnx we always cast the args (start, end, step) to desired dtype before computing size. 
On master, 
`
>>> import torch
>>> import numpy as np
>>> np.arange(1, 5.5, 1.5, dtype=np.int32)
array([1, 2, 3], dtype=int32)
>>> np.arange(1, 5.5, 1.5, dtype=np.int64)
array([1, 2, 3])
>>> torch.arange(1, 5.5, 1.5, dtype=torch.int32)
tensor([1, 2, 3], dtype=torch.int32)
>>> torch.arange(1, 5.5, 1.5, dtype=torch.int64)
tensor([1, 2, 3, 4])
`
This PR will change the last one to 
`
>>> torch.arange(1, 5.5, 1.5, dtype=torch.int64)
tensor([1, 2, 3])
`
but for onnx, 
'
>>> onnx.arange(1, 5.5, 1.5, dtype=torch.int32)
tensor([1, 2, 3, 4], dtype=torch.int32)
>>> onnx.arange(1, 5.5, 1.5, dtype=torch.int64)
tensor([1, 2, 3, 4])
`
the failed unittest compare torch and onnx with int64, so it works on master but fails with this PR. We may need to change the implementation in torch.onnx to make it consistent.

yuguo68 committed:
Update on "fix two corner cases of torch.arange"


trying to fix https://github.com/pytorch/pytorch/issues/80589, see the two corner cases in the issue. 
added the two cases in unit tests and add device to the tests. 

[ghstack-poisoned]
